### PR TITLE
SI-7377 Fix retypechecking of patterns on case companion alias

### PR DESCRIPTION
**Pull request by retronym**
Some ancient code in Typers switches from PATTERNmode to
EXPRmode when encountering `stableF(...)`. It just
typechecks `stableF` and discards the arguments.

To the best of Martin's recollection, this has something
to do with the need to typecheck patterns rather late in
the compiler, after `a.b` had been translated to `a.b()`
in `Uncurry`. I'm not able to motivate this with tests
using `-Xoldpatmat`; was there ever an even older pattern
matcher that ran _after_ uncurry?

What changed in 2.10.1 to expose this wrinkle? dfbaaa17
fixed `TypeTree.copyAttrs` to copy the original tree.
During the descent of `ResetAttrs`, sub-trees are duplicated
before begin further transformed. Duplicating the `Match`
in 2.10.0 would forget that the original tree of:

```
pat = (a: Int)Foo(_)
      `----------`
           `- TypeTree((a: Int)Foo), with original Select(..., "FooAlias")
```

The retypechecking would operate on the `MethodType`, rather
than the `Select`, which was not considered a stable
application.

For 2.10.x, I've just tightened up the condition to only
hit this if `args` is empty. I'm almost certain that the code
can be removed altogether, and I'll do that when this is merged
to master.

Review by @paulp

### CODE REVIEW

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 1]
```diff
@@ -4532,8 +4532,13 @@ trait Typers extends Modes with Adaptations with Tags {
 
       def normalTypedApply(tree: Tree, fun: Tree, args: List[Tree]) = {
         val stableApplication = (fun.symbol ne null) && fun.symbol.isMethod && fun.symbol.isStable
-        if (stableApplication && isPatternMode) {
+        if (args.isEmpty && stableApplication && isPatternMode) {
           // treat stable function applications f() as expressions.
+          //
+          // [JZ] According to Martin, thiw is relateed to the old pattern matcher, which
```
Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: "this is related"?

Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: um... yes.

Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: I believe the CPS plugin would typecheck patterns after uncurry.

On Wed, Apr 17, 2013 at 4:05 AM, Jason Zaugg notifications@github.comwrote:

> In src/compiler/scala/tools/nsc/typechecker/Typers.scala:
> 
> > ```
> >        // treat stable function applications f() as expressions.
> > ```
> > -          //
> > -          // [JZ] According to Martin, thiw is relateed to the old pattern matcher, which
> 
> um... yes.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/scala/scala/pull/2402/files#r3830547
> .

## 

See you at Scala Days http://scaladays.org (June 10–12, NYC)!

**Comment by paulp** [comment]
Github is driving me nuts. I received a comment from adriaan in email, but at no point has it been visible to me anywhere in the web interface. I guess he might have deleted it, but it didn't seem that sort of comment.

**Comment by paulp** [comment]
LGTM, especially the part where we remove it.

**Comment by retronym** [comment]
His comment is "I believe the CPS plugin would typecheck patterns after uncurry.", which is visible to me be expanding the outdated diff.

**Comment by paulp** [comment]
Oh, it says "xeno-by discussed an outdated diff." I guess that means "xeno-by and some unknown quantity of his discussion buddies."

**Pull request merged**